ext/openssl/ossl_pkey_rsa.c

```diff
@@ -628,6 +628,173 @@ ossl_rsa_to_public_key(VALUE self)
     return obj;
 }
 
+#if (OPENSSL_VERSION_NUMBER >= 0x10000000)
+/*
+ * call-seq:
+ *    rsa.sign_pss(digest, data) -> String
+ *
+ * Probabilistic Signature Scheme for RSA sign().
+ *
+ * To sign the +String+ +data+, +digest+ must be provided.
+ * +digest+ must be a OpenSSL::Digest instance or +String+ like "SHA1".
+ * The return value is again a +String+ containing the signature.
+ * A PKeyError is raised should errors occur.
+ * Any previous state of the +Digest+ instance is irrelevant to the signature
+ * outcome, the digest instance is reset to its initial state during the
+ * operation.
+ *
+ * == Example
+ *   data = 'Sign me!'
+ *   digest = OpenSSL::Digest::SHA256.new
+ *   pkey = OpenSSL::PKey::RSA.new(2048)
+ *   signature = pkey.sign_pss(digest, data)
+ */
+static VALUE
+ossl_rsa_sign_pss(VALUE self, VALUE digest, VALUE data)
+{
+    EVP_PKEY *pkey;
+    EVP_PKEY_CTX *pkey_ctx;
+    const EVP_MD *md;
+    EVP_MD_CTX *md_ctx;
+    size_t buf_len;
+    VALUE signature;
+    int error;
+
+    error = 0;
+
+    pkey = GetPrivPKeyPtr(self);
+    md = GetDigestPtr(digest);
+    StringValue(data);
+    signature = rb_str_new(0, EVP_PKEY_size(pkey));
+
+    md_ctx = EVP_MD_CTX_new();
+    if (!md_ctx)
+        ossl_raise(ePKeyError, "EVP_MD_CTX_new");
+
+    pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    if (!pkey_ctx) {
+        error = 1;
+        goto err;
+    }
+
+    if (EVP_DigestSignInit(md_ctx, &pkey_ctx, md, NULL, pkey) != 1) {
+        error = 2;
```

Comment:
The value of `error` is only used to determine if an error happens or not. I think this can be made more simple.

~~~c
    if (EVP_abc() != 1) {
	goto err;
    }

    if (EVP_def() != 1) {
	goto err;
    }

    EVP_MD_CTX_free(...);
    return signature;

err:
    EVP_MD_CTX_free(...);
    ossl_raise(...)
~~~